[SPARK-50675][SQL] Table and view level collations support

dejankrak-db opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'common/utils/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?

This change introduces table and view level collations support in Spark SQL, allowing CREATE TABLE, ALTER TABLE and CREATE VIEW commands to specify DEFAULT COLLATION to be used. For CREATE commands, this refers to all the underlying columns added as part of the table/view creation. For ALTER TABLE command, this refers to only newly created columns in the future, whereas existing ones are not affected, i.e. their collation remains the same.

The PR has been modelled after the original changes made by @stefankandic in https://github.com/apache/spark/pull/48090, with this PR covering table and view-level collations, whereas a follow up PR will be made covering schema-level collations.

This PR is adding/extending the corresponding DDL commands for specifying table/view level collation, whereas a follow up PR will be created separately to leverage the table/view collation in order to determine default collations for input queries of DML commands.

### Why are the changes needed?

From our internal users feedback, many people would like to be able to specify collation for their objects, instead of each individual columns. This change adds support for table and view level collations, whereas subsequent changes will add support for other objects such as schema-level collations.

### Does this PR introduce _any_ user-facing change?

The change follows the agreed additions in syntax for collation support.

The following syntax is now supported (**bold** parts denote additions):

{ { [CREATE OR] REPLACE TABLE | CREATE [EXTERNAL] TABLE [ IF NOT EXISTS ] }
  table_name
  [ table_specification ]
  [ USING data_source ]
  [ table_clauses ]
  [ AS query ] }

table_specification
  ( { column_identifier column_type [ column_properties ] ] } [, ...]
    [ , table_constraint ] [...] )

table_clauses
  { OPTIONS clause |
    PARTITIONED BY clause |
    CLUSTER BY clause |
    clustered_by_clause |
    LOCATION path [ WITH ( CREDENTIAL credential_name ) ] |
    COMMENT table_comment |
    TBLPROPERTIES clause |
    **DEFAULT COLLATION table_collation_name |**
    WITH { ROW FILTER clause } } [...]


CREATE [ OR REPLACE ] [ TEMPORARY ] VIEW [ IF NOT EXISTS ] view_name
    [ column_list ]
    [ schema_binding |
      COMMENT view_comment |
      TBLPROPERTIES clause |
      **DEFAULT COLLATION collation_name** ] [...]
    AS query


ALTER TABLE table_name
   { ADD COLUMN clause |
     ALTER COLUMN clause |
     DROP COLUMN clause |
     RENAME COLUMN clause |
     **DEFAULT COLLATION clause | …**
   }

### How was this patch tested?

Tests for the new syntax/functionality were added as part of the change. Also, some of the existing tests were extended/amended to cover the new DEFAULT COLLATION for table/view objects.

### Was this patch authored or co-authored using generative AI tooling?

No

dejankrak-db committed:
Initial implementation of table and view level collations

dejankrak-db committed:
Merge branch 'apache:master' into object-level-collations

dejankrak-db committed:
Additional AstBuilder handling for collations

dejankrak-db committed:
Fix AlterTableSetPropertiesCommand

dejankrak-db committed:
Adding tests

dejankrak-db committed:
Fix test failures

dejankrak-db committed:
Merge branch 'apache:master' into object-level-collations

dejankrak-db committed:
Additional fixes

dejankrak-db committed:
Fix test

dejankrak-db committed:
Merge branch 'apache:master' into object-level-collations

dejankrak-db commented:
Adding @cloud-fan, @stefankandic, @stevomitric, @uros-db - please take a look when you find some time, thanks!

dejankrak-db committed:
Merge branch 'apache:master' into object-level-collations

dejankrak-db committed:
Merge branch 'apache:master' into object-level-collations

dejankrak-db commented:
@stefankandic, @stevomitric, please take a look at the PR as discussed and let me know if you see any issues, thanks!

stefankandic reviewed (approved):
LGTM!

uros-db commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/DDLParserSuite.scala:
can we add at least some test cases for the UTF8 family of collations? UTF8_BINARY, UTF8_LCASE

also RTRIM versions...

dejankrak-db replied:
Sure, makes sense - I've extended the newly added DDLParserSuite test cases for creating/replacing/altering table to cover multiple different collations, including UTF8 family, UNICODE family (including RTRIM), as well as a couple of SR collations. Thanks for the suggestion!

dejankrak-db committed:
Merge branch 'apache:master' into object-level-collations

dejankrak-db committed:
Extending DDLParserSuite tests with additional UTF8 and RTRIM collations

cloud-fan commented on sql/api/src/main/scala/org/apache/spark/sql/catalyst/parser/DataTypeAstBuilder.scala:
```diff
@@ -232,6 +232,16 @@ class DataTypeAstBuilder extends SqlBaseParserBaseVisitor[AnyRef] {
+    ctx.asScala.headOption.map(visitCollationSpec)
+  }
+
+  override def visitCollationSpec(ctx: CollationSpecContext): String = withOrigin(ctx) {
```
How are these two methods related to `DataTypeAstBuilder`? They are for Spark DDL commands and should be put in `AstBuilder` I think, the same as other CREATE TABLE/VIEW clauses.

dejankrak-db replied:
Makes sense, moved these methods to AstBuilder.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:
```diff
@@ -119,6 +119,13 @@ class ResolveSessionCatalog(val catalogManager: CatalogManager)
         clusterBySpecOpt.getOrElse(ClusterBySpec(Nil)), conf.resolver))
       AlterTableSetPropertiesCommand(table.catalogTable.identifier, prop, isView = false)
 
+    case AlterTableCollation(ResolvedTable(catalog, _, table: V1Table, _), collation)
```
Since this is a new feature, I think we can directly go with the v2 command. i.e. do not match this command here, and eventually Spark will plan it as `AlterTableExec`

dejankrak-db replied:
@cloud-fan, actually this exposed a bug in v2AlterTableCommands implementation of AlterTableCollation(), which was not setting the collation properly for existing tables. I have changed this code to use setProperty instead (similar to how other table catalog props are altered), and added a test case in v2/DescribeTableSuite to ensure that this path now works correctly.

cloud-fan commented:
To confirm: this PR is adding/extending the related DDL commands for specifying table/view level collation, and we will open a followup PR to leverage the table/view collation to determine default collations for input queries of DML commands?

dejankrak-db committed:
Merge branch 'apache:master' into object-level-collations

dejankrak-db committed:
Refactoring parts of the code to address PR comments

dejankrak-db commented:
> To confirm: this PR is adding/extending the related DDL commands for specifying table/view level collation, and we will open a followup PR to leverage the table/view collation to determine default collations for input queries of DML commands later?

@cloud-fan, that is correct, I've also updated the PR description with this info to avoid any confusion, thanks for pointing this out! I have addressed all your comments so far and refactored the code accordingly, please let me know if you have any further comments/concerns, or the PR can be merged, thanks!

dejankrak-db committed:
Fixing AlterTableCollation command to use table property

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala:
```diff
@@ -304,12 +304,13 @@ case class AlterTableSetPropertiesCommand(
       properties = table.properties ++ properties,
-      comment = properties.get(TableCatalog.PROP_COMMENT).orElse(table.comment))
+      comment = properties.get(TableCatalog.PROP_COMMENT).orElse(table.comment),
+      collation = properties.get(TableCatalog.PROP_COLLATION).orElse(table.collation))
```
how can we hit this code path? People use SET TBLPROPERTIES command and explicitly set the `collation` property?

dejankrak-db replied:
Actually, @cloud-fan, you are right, because we raise an error in AstBuilder.cleanTableProperties() when someone attempts to change collation through SET/UNSET TBLPROPERTIES, so no changes to the collation should be possible here or below in AlterTableUnsetPropertiesCommand - hence, I've reverted the changes in this file. Good catch!

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/execution/command/v2/DescribeTableSuite.scala:
```diff
@@ -213,4 +213,29 @@ class DescribeTableSuite extends command.DescribeTableSuiteBase
     }
   }
+
+  Seq(true, false).foreach { hasCollations =>
```
what's the difference between this one and the one in `v1/DescribeTableSuite`? Can we move it up to the parent suite?

dejankrak-db replied:
Moved the tests to DescribeTableSuiteBase suite, to avoid code duplication, whereas these test cases are now run as part of both v1 and v2 suites for complete coverage.

cloud-fan commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HiveClientSuite.scala:
```diff
@@ -316,6 +335,17 @@ class HiveClientSuite(version: String) extends HiveVersionSuite(version) {
     assert(!client.tableExists("temporary", "tgt"))
   }
 
+  test("alterTable - change collation") {
```
is it the same with the test `create/alter table with collations`?

dejankrak-db replied:
Right, this test case covers a subset of that one (just the alter table default collation part), so I have removed this test case.

dejankrak-db committed:
Addressing further comments, refactoring parts of code/tests

dejankrak-db committed:
Merge branch 'apache:master' into object-level-collations

dejankrak-db committed:
Minor indendtation fix

dejankrak-db commented:
@cloud-fan, thanks for the review and great suggestions! I have addressed them all and changed/refactored the code accordingly, I believe the PR should be ready for merge, please let me know if you have any additional comments, thanks a lot!

dejankrak-db commented:
Created https://issues.apache.org/jira/browse/SPARK-50522 and linked PR to it

cloud-fan commented:
thanks, merging to master!
